### DISCUSSION TITLE
Concept of achievement's maintainer

### DISCUSSION THREAD

**Discussion by meleu** [RetroAchievements/RAWeb#803] [Ideas]
## UPDATE

**jump directly to the discussion [starting here](https://github.com/RetroAchievements/RAWeb/issues/520#issuecomment-849902103)**

### original post

In reply to @televandalist's comment here: https://github.com/RetroAchievements/RAWeb/pull/519#issuecomment-727282927

I'm under the impression that what @televandalist wants is to have one user to be the care taker of the achievement (the person responsible to solve the achievement's tickets), and allow volunteers to have such role, even if they are not the original author.

IMO changing the authorship is not a good way to achieve that. A better approach could be to implement the concept of `maintainer`.

How it would be managed is still something that needs more discussions.

Some questions to be answered:
- should achievement creators be allowed to set this role to themselves?
- should achievement creators be allowed to unset this role from themselves?
- if maintainers have open tickets should they be allowed to work on new sets?

**Comment by televandalist** [comment]
Thanks

A couple responses to @luchaos in #519 
> I'm not sure I follow :) There is no "loss of authorship" - nor is it planned?

I know it isn't planned; I'm just throwing ideas out there. The "loss of authorship" would be temporary and would be reinstated once there's multiple authors per achievement, which is why I mentioned keeping a paper trail. It could be as simple as this:
![987987987](https://user-images.githubusercontent.com/45054151/99161282-595b6f80-26b6-11eb-9163-bd16b3b59217.PNG)
Once multiple authors are a thing, the list could be referenced to reinstate the original authorship if need be. 

> What should happen down the line is that the "single responsibility and ownership" comes to an end - where anyone who feels like it can contribute to an achievement's improvement and not only the original author is bothered with ticket requests.

I agree with this, but keep in mind I'm only thinking in V1 terms because it's the only reference I have. 

-------------------

@meleu 

> I'm under the impression that what @televandalist wants is to have one user to be the care taker of the achievement (the person responsible to solve the achievement's tickets), and allow volunteers to have such role, even if they are not the original author.

Correct, but only "for now" - at least until multiple authors are a thing, like stated previously. My main objective here is a QA move to basically motivate *active devs* to light a fire under their own asses by volunteering to take full responsibility for abandoned work. If the original author's account was deleted, banned, or they just haven't logged in forever and became inactive, the end result is the same: problems with their achievements are low-priority by default, regardless of how popular the set is. Aside from a small handful of devs who periodically work to fix these issues (❤️), these achievements typically sit broken and only have a chance of being resolved during dev events. Imagine as a player, reading [How to contribute if you are not a developer](http://docs.retroachievements.org/How-to-contribute-if-you-are-not-a-developer/) and going all in with ticket submissions.... only to not have them addressed for sometimes 2+ years. Even if the issue is something as dumb/simple as changing "=" to ">="... Anyway, transferring authorship to volunteering active devs would turn those issues into high-priority. Last time I checked, there were 1400+ open tickets, over 90% of which were for achievements by inactive devs. It's a huge problem and with the userbase continuing to increase, it's going to become more evident. 

> IMO changing the authorship is not a good way to achieve that. A better approach could be to implement the concept of maintainer.

I gotta head out for an hour or so, but I'll come back to address this as well as the questions. Will also reference my original proposal.

**Comment by televandalist** [comment]
@meleu I took longer than expected, so new post it is.

I think one of the reasons the reauthoring proposal seem so drastic is that at face value, it seems like a way for some devs to quickly hoard authorship, which is just an accident waiting to happen. Someone wanting to adopt achievements "just because" wouldn't fly. In #519 , I mentioned that there would need to be specific guidelines and criteria involved. These would be to ensure that the dev adopting the achievement(s) *really* wants to be responsible for them. I'll dive into what I had in mind:

*keep in mind I'm still just throwing ideas around so there's still holes in this... I'd love additional ideas, suggestions, arguments, etc.*

Before achievement(s) can be eligible for adoption, one of the following must be true:
- The original author must have a profile deletion request granted.
- The original author has been banned without an appeal for at least [uncertain duration]
- The original author has been marked as inactive for at least [uncertain duration]
- The original author has not logged in for at least [uncertain duration]
- The original author relinquishes responsibility to maintain the achievement(s); therefore, disowning it/them. (this answers `should achievement creators be allowed to unset this role from themselves?`)

So I can wait until 2020-12-19 and adopt all 5415 of Salsa's cheevos to become Tele EX+? Absolutely not. 

In order to adopt an eligible achievement:
- The adopter must have resolved a least one **legitimate** ticket for that achievement, prior to the request.
- The adopter must be free of tickets at time of request.
- The adopter must have a positive (green) Resolved Ticket Karma.
- As an alternative to resolving tickets, achievements modified by the adopter during a revision may be eligible. [Would need further discussion] 

In the extremely unlikely scenario that two devs want to adopt the same achievement(s) and can't come to a compromise, the following could be taken into consideration:
- Have either of them been involved with a prior revision of the set? 
- Are either of them credited as an author for other achievements in the set? 


Notes:

- In some cases, one achievement is reported, but several other achievements in the set have the same issues, whether they're mentioned or not. These achievements should also be eligible. Logic snapshots would likely be needed. 
- Another thing that would need to be discussed is if a dev ends up maintaining an entire set, they would become the "sole author" and would no longer require approval to do revisions, rescores, etc. for that set.

---------

Okay, so, `maintainers` - I think I would need to know more about what you feel the maintainer role would entail before I comment too much on it, but I'll try. 

> IMO changing the authorship is not a good way to achieve that. A better approach could be to implement the concept of maintainer.

How would this be incorporated into V1? The way I'm picture reauthoring is a simple `this dev` -> `that dev` change and wouldn't affect anything else on the site, that I know of. Dev stats might get a little screwy, though. 🤔 

> should achievement creators be allowed to set this role to themselves?

How I'm picturing it, achievement creators would also be the maintainers by default, for their work anyway. For others' work, I feel like it would be close to the criteria I mentioned earlier in this post.

> should achievement creators be allowed to unset this role from themselves?

For their own achievements, no, not without also unsetting the authorship. For others' work, I don't see why not.

> if maintainers have open tickets should they be allowed to work on new sets?

I think it should be viewed the same as being the author. Otherwise it would kinda defeat the purpose. So I would have to say no, at least not while the 0 ticket rule is in place.

-----

All in all, I don't think either reauthoring or a maintainer role would be difficult to manage. Lots of explaining and discussions to be had, though 💪

**Comment by HolyShinx** [comment]
I really love reading that concept tele, there is one issue, which is "The adopter must have resolved a least one legitimate ticket for that achievement, prior to the request." which means to adopt the set every achievement must have gotten a ticket, and it must be resolved by the adopter, what if the game just isn't popular?

**Comment by televandalist** [comment]
> I really love reading that concept tele, there is one issue, which is "The adopter must have resolved a least one legitimate ticket for that achievement, prior to the request." which means to adopt the set every achievement must have gotten a ticket, and it must be resolved by the adopter, what if the game just isn't popular?

I kind of touched on this already: `In some cases, one achievement is reported, but several other achievements in the set have the same issues, whether they're mentioned or not. These achievements should also be eligible. Logic snapshots would likely be needed.` Same thing could be said of unpopular sets. If the problems exist, show it and start the paper trail. 

Outside of major revisions and repairing completely busted sets, I don't think there would be a need to adopt an entire set. That seems like it'd be more along the lines of adopting something for the sole purpose of just slapping a name on it, which is the main situation I want to avoid. Let's say there's an abandoned set that has 50 achievements. 45 of them work perfectly, but 5 of them are completely busted. There's no need to adopt the 45 achievements that work. 

Personally, I like the idea of a potential adopter fixing the achievements and resolving the tickets before requesting to adopt. Putting work into it beforehand shows they care enough about them to adopt them. Plus, Resolved Ticket Karma if they didn't open the tickets themselves😏

**Comment by Alfex3** [comment]
> The adopter must have a positive (green) Resolved Ticket Karma.

So if someone makes too many mistakes in their achievements, even if the achievements work for everyone in the end, they won't be able to adopt a set? I think i got my ticket karma by resolving tickets of games I added region support to. If the concept of maintainer had been implemented earlier, my ticket karma would be negative.

**Comment by televandalist** [comment]
Nothing has been implemented. This is all just spitballing. Though I would like to see ticket karma go beyond a statistic and get a practical use.

**Comment by meleu** [comment]
I'd like to start working on this.

Before implementing I'm going expose my current plan and request input from the QA Team members.

## proposed change

1. Create a new column named `Maintainer` in the `GameData` **and** `Achievements` tables
    - `Maintainer` in `GameData` actually meaning "achivement set maintainer"
    - `Maintainer` in `Achievements` meaning "individual achievement maintainer"
    - in any case, it accepts `null`

2. From a website user point of view:
    - achievement set maintainer would be visible in the game's page (right below the Authors list).
    - individual achievement maintainer would be visible in the achievement's page (right below the "Created by" line)
    - in both cases, it would editable by admins in "Dev (Click to show)" menu
    - maintainers should be visible in the ticket's page (@SlashTangent's suggestion)

**NOTE**: at first it would be a field editable only by admins, but we can add some automation later (see "discussions" at the end of this post).


## use cases

- It would allow volunteers to adopt "abandoned" sets.

- Maintainers would receive all ticket notifications for their maintained set.

- In the user's page, it would have a link to all sets maintained by that user (this may sound like a "vanity" thing, but I'm sure it'll be useful)

- **Possible expansion of functionality**:
    - If a game doesn't have any achievements, the "game's maintainer" concept can be used for claims (interesting idea brought by @luchaos).


## discussion that will possibly emerge

- Will maintainers be allowed to claim new games if their maintained games have tickets?
    - my opinion: not really sure... that would surely influence the number of volunteers to be maintainers

- sole developers should automatically be set as the game's maintainer?
    - my opinion: yes

- should the last person who edited an achievement be automatically set as its maintainer?
    - my opinion: first reaction: "yes, of course!". But it's not that simple. Sometimes people just want to update the badge or fix grammar mistakes... So: currently **no**.

**Comment by StingX2** [comment]
_Will maintainers be allowed to claim new games if their maintained games have tickets?_

This part can undermine our claim system and cause extra issues for QATeam

**Comment by televandalist** [comment]
*Will maintainers be allowed to claim new games if their maintained games have tickets?*

I'd say yes. If it gets out of hand, then change the maintainer to someone more willing. Would need a bigger discussion, though. 

*sole developers should automatically be set as the game's maintainer?*

Definitely

*should the last person who edited an achievement be automatically set as its maintainer?*

Hell no. Sting and I would be made maintainers for like half the site after all those rescores.

**Comment by Alfex3** [comment]
_Will maintainers be allowed to claim new games if their maintained games have tickets?_
If they are, what's the point of the original rule "you must be free of tickets before claiming a game"? 
_should the last person who edited an achievement be automatically set as its maintainer?_
I'd say, if the code was changed, yes; if only the badge, title, description or score was changed, no.
